### PR TITLE
GH-37720: [Format][Docs][FlightSQL] Document stateless prepared statements

### PR DESCRIPTION
**Pull request by erratic-pattern**
documents changes for stateless management of FlightSQL prepared statement handles based on the design proposal described in apache/arrow#37720

* GitHub Issue: #37720

PRs for language implementations:
* Rust: apache/arrow-rs#5433
* Go: apache/arrow#40311

Mailing list discussion: https://lists.apache.org/thread/3kb82ypx99q96g84qv555l6x8r0bppyq

### CODE REVIEW

File: format/FlightSql.proto [hunk 1]
```diff
@@ -1797,6 +1797,15 @@ message DoPutUpdateResult {
   int64 record_count = 1;
 }
 
+// An optional response returned when `DoPut` is called with `CommandPreparedStatementQuery`
+message DoPutPreparedStatementResult {
+  option (experimental) = true;
+
+  // (potentially updated) opaque handle for the prepared statement on the server.
+  // All subsequent requests for his prepared statement must use this new handle, if specified
```
Comment (lidavidm) [inline format/FlightSql.proto:1805]:
```suggestion
  // All subsequent requests for this prepared statement must use this new handle, if specified
```

**Review by lidavidm — APPROVED** [review]
Do we need to discuss backwards compatibility at all? (Presumably it's on the server to detect that the client didn't use the updated handle?)

File: format/FlightSql.proto (hunk 1 above)
Comment (erratic-pattern) [reply format/FlightSql.proto:1805]: > Do we need to discuss backwards compatibility at all? (Presumably it's on the server to detect that the client didn't use the updated handle?)

The only backwards compatibility safeguard in the current proposal is that the client must assume the returned handle could be missing if the server is following the previous spec. I am open to other suggestions.

What should happen here depends on whether the server is stateful or stateless.

*Stateful*: This should be inherently backwards and forwards compatible by design. There is no need for an updated handle here, because the server is already managing the necessary state. The server can simply execute the statement with the original handle, or any updated handle, since it should have all the necessary state for the prepared statement. 

*Stateless*: There is no way to properly support prepared statements if the client does not return the updated handle. There is no backwards compatibility here, and so the only thing to do here is to fail gracefully

It may be worthwhile to document this somewhere, but I'm not sure the best place for it.

It would be nice if there could be a way for server's to know whether or not a client supports the updated protocol, as that would provide a more straightforward path for servers to provide graceful failure. But I'm not sure of the best way to do that. There may be implementation-specific tricks involving checking a log or a timestamp embedded in the handle to determine if a handle was updated but I think that's out of the scope of the spec

File: docs/source/format/FlightSql.rst [hunk 2]
```diff
@@ -141,6 +141,9 @@ the ``type`` should be ``ClosePreparedStatement``).
     Execute a previously created prepared statement and get the results.
 
     When used with DoPut: binds parameter values to the prepared statement.
+    The server may optionally respond with an updated handle. The client
+    should use this updated handle for all subsequent requests for this
+    prepared statement.
```
Comment (alamb) [inline docs/source/format/FlightSql.rst:146]:
```suggestion
    should use this updated handle for all subsequent requests for this
    prepared statement. The updated handle allows implementing query 
    parameters with stateless services. Note that the server is responsible 
    for detecting the case where the client does not use the updated handle on 
    subsequent requests (older clients may ignore this field) and responding 
    appropriately.
```

File: format/FlightSql.proto [hunk 3]
```diff
@@ -1797,6 +1797,15 @@ message DoPutUpdateResult {
   int64 record_count = 1;
 }
 
+// An optional response returned when `DoPut` is called with `CommandPreparedStatementQuery`
+message DoPutPreparedStatementResult {
+  option (experimental) = true;
+
+  // (potentially updated) opaque handle for the prepared statement on the server.
+  // All subsequent requests for this prepared statement must use this new handle, if specified
```
Comment (alamb) [inline format/FlightSql.proto:1805]: I think it might help to give a hint here about what this field is used for

```suggestion
  // All subsequent requests for this prepared statement must use this new handle, if specified.
  // The updated handle allows implementing query parameters with stateless services
  // as described in https://github.com/apache/arrow/issues/37720
```

**Review by alamb — APPROVED** [review]
Thank you very much @erratic-pattern  - this looks great

> Do we need to discuss backwards compatibility at all? (Presumably it's on the server to detect that the client didn't use the updated handle?)

Thank you @lidavidm  that is a great idea. I left some potential text for this

This also made me wonder if we do (or should) version the FlightSQL protocol (so for example, we could describe "this client supports version 1.2" or something)

We should wait until a formal vote is held on the mailing list to merge this PR but from my perspective this PR is good to go

File: format/FlightSql.proto (hunk 3 above)
Comment (erratic-pattern) [reply format/FlightSql.proto]: > It would be good to copy/summarize the justification here, if we think it's worth keeping, instead of/along with referencing the ticket

I've updated the docs to provide some background on this change, as well as explicitly documenting legacy behavior. also made the updated handle explicitly `optional` to help indicate that the older stateful pattern can still be supported even in up-to-date clients and servers.

File: docs/source/format/FlightSql.rst [hunk 4]
```diff
@@ -141,6 +141,13 @@ the ``type`` should be ``ClosePreparedStatement``).
     Execute a previously created prepared statement and get the results.
 
     When used with DoPut: binds parameter values to the prepared statement.
+    The server may optionally respond with an updated handle. The client
+    should use this updated handle for all subsequent requests for this
+    prepared statement. The updated handle allows implementing query 
+    parameters with stateless services. Note that the server is responsible
```
Comment (pitrou) [inline docs/source/format/FlightSql.rst:147]: This is very difficult to understand. How does an updated handle allow implementing query parameters? This should be more descriptive (does the handle embody the bound parameters?).

Also, "the client should use this updated handle" means the client cannot bind the original prepared statement to other parameters?

File: format/FlightSql.proto [hunk 5]
```diff
@@ -1797,6 +1797,27 @@ message DoPutUpdateResult {
...
+ */
+message DoPutPreparedStatementResult {
+  option (experimental) = true;
+
+  // Represents a (potentially updated) opaque handle for the prepared statement on the server.
+  // Because the handle could potentially be updated, any previous handles for this prepared
+  // statement should be considered invalid, and all subsequent requests for this prepared
+  // statement must use this new handle, if specified.
+  // The updated handle allows implementing query parameters with stateless services
+  // as described in https://github.com/apache/arrow/issues/37720.
```
Comment (pitrou) [inline format/FlightSql.proto:1814]: We should not point to GH issue discussions as an explanation of why this exists. The spec should ideally be self-contained.

File: format/FlightSql.proto [hunk 6]
```diff
@@ -1797,6 +1797,27 @@ message DoPutUpdateResult {
...
+ * this command, and that behavior should still be supported by clients. See documentation
+ * of individual fields for more details on expected client behavior in this case. 
+ */
+message DoPutPreparedStatementResult {
+  option (experimental) = true;
+
+  // Represents a (potentially updated) opaque handle for the prepared statement on the server.
+  // Because the handle could potentially be updated, any previous handles for this prepared
+  // statement should be considered invalid, and all subsequent requests for this prepared
+  // statement must use this new handle, if specified.
```
Comment (pitrou) [inline format/FlightSql.proto:1812]: Why "if specified"?

File: docs/source/format/FlightSql.rst [hunk 7]
```diff
@@ -141,6 +141,13 @@ the ``type`` should be ``ClosePreparedStatement``).
     Execute a previously created prepared statement and get the results.
 
     When used with DoPut: binds parameter values to the prepared statement.
+    The server may optionally respond with an updated handle. The client
+    should use this updated handle for all subsequent requests for this
+    prepared statement. The updated handle allows implementing query 
+    parameters with stateless services. Note that the server is responsible 
+    for detecting the case where the client does not use the updated handle on 
+    subsequent requests (older clients may ignore this field) and responding 
+    appropriately.
```
Comment (pitrou) [inline docs/source/format/FlightSql.rst:150]: What does "appropriately" imply? Return an error?

File: format/FlightSql.proto (hunk 1 above)
Comment (alamb) [reply format/FlightSql.proto]: In general, I think we should have a larger discussion about how to communicate / handle compatibility in the FlightSQL protocol -- I filed https://github.com/apache/arrow/issues/40424 to discuss the over arching issue, so hopefully we can focus discussions on this ticket to this particular proposal

File: docs/source/format/FlightSql.rst (hunk 4 above)
Comment (alamb) [reply docs/source/format/FlightSql.rst:147]: > . How does an updated handle allow implementing query parameters? This should be more descriptive (does the handle embody the bound parameters?).

Maybe we can say something like this:

> Optionally updating the handle allows the client to supply all state required to execute the query in an `ActionPreparedStatementExecute` message. For example, stateless servers can encode the bound parameter values into the new handle, and the client will send that new handle with parameters back to the server. 


The problem is  as currently written, the `CommandPrepareStatementQuery` message contains parameter values, but does not return anything to the client prior to the client sending `ActionPreparedStatementExecute`. Thus the only way to implement prepared statements with bind parameters today is to store the value of the parameters on the server between the call to `CommandPrepareStatementQuery`  and `ActionPreparedStatementExecute`.


> Also, "the client should use this updated handle" means the client cannot bind the original prepared statement to other parameters?

That is an excellent question. I think the intent is that the handle could be updated with new parameter values. I agree as worded this is confusing and should be clarified 

>  Note that a handle returned from a DoPut call with `CommandPreparedStatementQuery` can itself be passed to a subsequent DoPut call with `CommandPreparedStatementQuery` to bind a new set of parameters. The subsequent call itself may return an updated handled which again should be used for subsequent requests.

**Review by alamb — COMMENTED** [review]
Thank you for your review and helpful comments @pitrou  -- I left some suggestions on how to clarify your questions. Let me know what you think @erratic-pattern

**Comment by joellubi** [comment]
Will we need to extend the schema of the `PutResult` message in order to encapsulate the `DoPutPreparedStatementResult` payload?

This is the spec right now:
```proto
service FlightService {
  rpc DoPut(stream FlightData) returns (stream PutResult) {}
}

message PutResult {
  bytes app_metadata = 1;
}
```

I don't think the response would be considered `app_metadata`.

**Comment by lidavidm** [comment]
The "app" in this case is Flight SQL. (Someday I would like to fuse Flight SQL/Flight RPC to avoid these odd distinctions, and so that we aren't multiplexing a dozen different calls into the same endpoint...)

**Comment by joellubi** [comment]
With that framing it makes sense, thanks @lidavidm.

File: format/FlightSql.proto [hunk 8]
```diff
@@ -1797,6 +1797,27 @@ message DoPutUpdateResult {
   int64 record_count = 1;
 }
 
+/* An *optional* response returned when `DoPut` is called with `CommandPreparedStatementQuery`.
+ *
+ * *Note on legacy behavior*: previous versions of the protocol did not return any result for
+ * this command, and that behavior should still be supported by clients. See documentation
+ * of individual fields for more details on expected client behavior in this case.
```
[the remark this answers is not in the document]
Comment (erratic-pattern) [reply format/FlightSql.proto:1804]: > Maybe we could phrase this in terms of what happens in the case where clients or servers don't support this new message

I wanted to avoid documenting information about individual fields in the message, since the message might one day contain many fields.

I think I could simplify the wording on this to just indicate that older servers dont use this, clients should support that, and just use the default blank values for fields in that case.

**Review by alamb — APPROVED** [review]
With the updated changes this PR looks good to me

**Review by alamb — APPROVED** [review]

**Comment by erratic-pattern** [comment]
This proposal was approved on the Arrow dev mailing list. [See thread here](https://lists.apache.org/thread/zoyv6d4fzco4h48b66wqwmfc6w31757x)

**Comment by alamb** [comment]
Thank you everyone for your feedback and participation in the process
